lint/mypy.bzl

```diff
@@ -0,0 +1,231 @@
+"""API for declaring a mypy lint aspect that visits py_library rules.
+
+Typical usage:
+
+```
+load("@aspect_rules_lint//lint:mypy.bzl", "mypy_aspect")
+
+mypy = mypy_aspect(
+    binary = "@@//:mypy",
+    configs = "@@//:pyproject.toml",
+)
+```
+
+Optional: enable it in `.bazelrc`:
+
+```
+# Add mypy type-check validation actions to py_library targets.
+# Disable for a particular build with --norun_validations
+build --aspects=//tools:lint.bzl%mypy
+```
+
+### Reporting
+
+Unlike most linters hosted in rules_lint, mypy produces only error semantics.
+That means that typecheck violations will input_depsets in failed build actions, rather than
+a report of warnings which can be handled in various ways.
+See https://github.com/aspect-build/rules_lint/blob/main/docs/linting.md
+
+However, it reports the violations using Bazel's Validation Actions feature, which means
+you can pass the `--norun_validations` flag to skip type-checking for a particular build.
+
+### Acknowledgements
+
+This code inspired from https://github.com/bazel-contrib/bazel-mypy-integration
+Thanks to the [contributors](https://github.com/bazel-contrib/bazel-mypy-integration/graphs/contributors)
+especially [Jonathon Belotti](https://github.com/thundergolfer) and [David Zbarsky](https://github.com/dzbarsky).
+
+### TODO
+
+- Allow configured typeshed repo, e.g. args.add("--custom-typeshed-dir", "external/my_typeshed")
+- Avoid invalidating caches whenever mypy.ini changes
+- Remote cache: bootstrap the stdlib since it will remain in cache, making other actions slightly faster
+- Later: Generate `.pyi` outputs: optimization to avoid as much invalidation
+  when only implementation bodies change, at the cost of an extra mypy action.
+"""
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+_MNEMONIC = "mypy"
+MypyInfo = provider(doc = "Python typechecking data", fields = {
+    "transitive_cache_map": "depset: transitive --cache-map json files produced by deps",
+})
+
+def mypy_info(direct, deps):
+    return MypyInfo(
+        transitive_cache_map = depset(
+            direct = direct,
+            transitive = [d[MypyInfo].transitive_cache_map for d in deps if MypyInfo in d],
+        ),
+    )
+
+def _extract_transitive_inputs(deps):
+    input_depsets = []
+    for dep in deps:
+        # Supply the --cache-map files from transitives
+        if MypyInfo in dep:
+            input_depsets.append(dep[MypyInfo].transitive_cache_map)
+
+        # NB: relies on PyInfo being a Bazel global symbol, one day we should be forced to load it
+        # TODO: maybe we can avoid passing .py source files when the cache-map files were found?
+        if PyInfo in dep:
```

Comment:
Should load through rules_python now as `incompatible_python_disallow_native_rules` might be enabled

Reply:
I agree, though I was trying to avoid the rules_python dependency. I think it's needed for provider symbols which is why this repo depends on rules_js too.

Reply:
Actually I can't enable the flag, seems like an upstream problem. Filed https://github.com/bazelbuild/rules_python/issues/1645

Reply:
added a TODO